Port v0.6 patches to current upstream kernel

chuckwolber opened the pull request:
Port the v0.6 patches to the current upstream Linux kernel 6.11-rc3. KUnit tests caused a kernel panic, so I am reasonably certain `LLVM_COV` needs to be disabled in additional places, but this should be a start.

chuckwolber committed:
Port v0.6 patches to current upstream kernel

Port the v0.6 patches to the current upstream Linux kernel 6.11-rc3 and
promote the result to v1.0. These patches will be developed as a v1.0
series until they are suitable for submission. At that point, they will
be merged from -next to the release branch.

Signed-off-by: Chuck Wolber <chuck.wolber@boeing.com>

whentojump commented:
Hi Chuck, thanks for the port!

> KUnit tests caused a kernel panic

Which specific tests were failing? Is it related to `CONFIG_KUNIT_FAULT_TEST`? Can you try disabling this option and running again?

whentojump commented:
In this repo we have been continuously testing with stable 5.15.y.  The actual patch development however happens in this tree https://github.com/xlab-uiuc/linux-cov/commits/v0.6/ whose base is 6.8.y. If you find it more convenient you may also work on that one. And it's pretty much appreciated!

I know between 5.15 and 6.8.y there would be some conflicts but they are relatively trivial. Between 6.8.y and mainline, one change (https://lore.kernel.org/all/20240506133544.2861555-1-masahiroy@kernel.org/, which removes many `GCOV_PROFILE := n` directives) might affect out patch `[RFC PATCH 2/3] kbuild, llvm-cov: disable instrumentation in odd or sensitive code`. It's still manageable though.

I can take care of all the rebasing stuffs. Let's first address the KUnit issue you mentioned.

chuckwolber commented:
@whentojump - Ok, sorry for the mixup. There was some confusion on my part as to how you are managing development. I submitted [a pull request ](https://github.com/xlab-uiuc/linux-mcdc/pull/12) that attempts to clarify the documentation. Feel free to make changes or reject the pull request altogether if you disagree.

Also, it is not completely clear to me how you are using branching, and I noticed that you recently renamed the default branch from `public-approved` to `llvm-trunk`. The wide variety of branches in `linux-mcdc` and `linux-cov` as well as the changing branch names makes it difficult to track what you are doing. If you could perhaps narrow things down a bit and shoot for stable and consistent naming, it will be easier to collaborate and make progress.

## Community Review
The v0.6 patchset applied to kernel `5.15.y` looks great and seems exceptionally stable to me. I also agree that [the change that removed](https://lore.kernel.org/all/20240506133544.2861555-1-masahiroy@kernel.org/) most of the `GCOV_PROFILE := n`  entries is pretty much the only thing that introduces instability between `6.8.x` and the head of the current mainline. 

If you feel strongly about getting RFC feedback from the community as soon as possible, I think it is fine if you want to send the current v0.6 patchset to the kernel developer community for review. The only things we would need to do to prepare are:

- Decide on the list of recipients.
- Update the cover page patch to be clear about which kernel the patches are tested on.
- Update the cover page patch to be clear about what remains to be done to get the patchset ready to merge to the head of the kernel mainline.

The downside of submitting a patchset targeted at a maintenance (`v5.15.y`) or unmaintained (`v6.8.y`) release tree is that busy kernel developers may choose to ignore it. But that is probably the worst case, and I do not think it will hurt you in the long run.

The patchset is ultimately going to need to work against the head of the mainline kernel to be ready for upstream submission, hence the reason for this pull request. But since we are getting close to LPC, your best bet may be to do the following:

1. Submit the v0.6 patchset for community RFC as stated above.
2. Create a `v1.0-draft` branch on `linux-cov` that is rebased every day against the Torvalds kernel.
3. Keep the updated set of v1.0 patches in a development branch (e.g. `llvm-trunk-next`) on this project (`linux-mcdc`).

This gives you a "known good" patchset that you can send out for community review while you are working on the v1.0 patchset against the latest mainline kernel. Once v1.0 is ready for submission to kernel maintainers, you can merge `llvm-trunk-next` to `llvm-trunk`.

This pull request could serve as the basis of the v1.0 patchset. I would only need to change the directory name from `v0.6.1` to `v1.0` and rebase my branch against a `llvm-trunk-next` branch once you create it.

whentojump commented:
Sorry for the confusion. Here's the clarification:

In short, for the patches themselves, you may focus on the emails, or https://github.com/xlab-uiuc/linux-cov/tree/v0.6. 

1. We develop the actual patch in https://github.com/xlab-uiuc/linux-cov/. Tags `v0.1` through `v0.6` are exactly what were sent in internal emails. Their base is 6.8.y (five months ago, when we started). And yes we *will* rebase the patches onto mainline eventually. The repo is *not* open (which is subject to Boeing approval).
2. https://github.com/xlab-uiuc/linux-mcdc is more or less a demo/advertisement/preview to the public before we can open other repositories (which is subject to Boeing approval). And to be less painful in automatic tests, patches are backported onto a fixed version 5.15.153.

Sorry I didn't particularly expect others to collaborate on the 2nd repo. From now on I will take full care of all the branches. 

> Community Review...

These are good advice and let's do that. I have created `llvm-trunk-next`. You are also given write access to this repo.

Thanks,
Wentao

whentojump commented:
cc @svleest

whentojump commented:
[`v1.0-draft`](https://github.com/xlab-uiuc/linux-cov/commits/v1.0-draft/) is uploaded. As Chuck suggested, this branch will be continuously rebased onto the latest tag in upstream. As of today it's `6.11-rc3`.

I've tested it locally. [CI runs](https://github.com/whentojump/linux-mcdc/actions) are also initiated in my fork. Once Chuck merges this PR in `llvm-trunk-next` I will also get my CI scripts in.

@chuckwolber Any update on the panic you mentioned in the original post? Does the kernel message look like the below before the panic? If so I think it's expected and can be eliminated by turning off `CONFIG_KUNIT_FAULT_TEST`.

```text
[    4.111299] ok 26 kunit_device
[    4.112655]     KTAP version 1
[    4.113192]     # Subtest: kunit_fault
[    4.113833]     # module: kunit_test
[    4.113837]     1..1
[    4.114920] BUG: kernel NULL pointer dereference, address: 0000000000000000
```

And what's your suggestions on the recipient list? (For this, let's maybe discuss in email)

chuckwolber commented:
Rebased this branch to `llvm-trunk-next`, moved the patches from `v0.6.1` to `v0.1`, and updated the commit message.

@whentojump - Going forward, in order to coordinate work in an orderly fashion, I recommend you make changes to this repository from a feature branch on your private clone, the same way I am doing. You can sync `llvm-trunk-next` to your private clone and create your feature branches from there. That gives us a common development branch to work.

When you wish to promote changes  from `llvm-trunk-next` to the release branch `llvm-trunk`, you can either create a pull request to sync all the changes, or do a traditional merge from the CLI if you want to be more granular and only promote specific commits.

chuckwolber commented:
> @chuckwolber Any update on the panic you mentioned in the original post? Does the kernel message look like the below before the panic? If so I think it's expected and can be eliminated by turning off `CONFIG_KUNIT_FAULT_TEST`.
> 
> ```
> [    4.111299] ok 26 kunit_device
> [    4.112655]     KTAP version 1
> [    4.113192]     # Subtest: kunit_fault
> [    4.113833]     # module: kunit_test
> [    4.113837]     1..1
> [    4.114920] BUG: kernel NULL pointer dereference, address: 0000000000000000
> ```

Yes, that is what I saw. I did not test to see if we get a kernel panic without the `llvm-cov` patches applied. If that happens to be the case then we can ignore the problem and omit `CONFIG_KUNIT_FAULT_TEST`. Otherwise, we should probably create an issue and get it resolved.

> And what's your suggestions on the recipient list? (For this, let's maybe discuss in email)

Yes, email discussion. I will reply to the proposed list you sent via email.

chuckwolber commented:
This was inadvertently merged to `llvm-trunk` instead of `llvm-trunk-next`.

Since there was an extra documentation commit on `llvm-trunk` that had not made it into `llvm-trunk-next`, I took the opportunity to cherry-pick the documentation commit over to `llvm-trunk-next`, and pick the commits from this PR to `llvm-trunk-next` as well. I rebased `llvm-trunk` to remove the commits from this PR and then I rebased `llvm-trunk-next` to `llvm-trunk` to ensure `llvm-trunk-next` properly branches from the current HEAD of `llvm-trunk`.

The end result is that `llvm-trunk-next` correctly branches from the HEAD commit on `llvm-trunk` and the commits from this PR only show up on `llvm-trunk-next`.

@whentojump - You will probably need to sync your private clone and do a `git pull --rebase` on `llvm-trunk` and `llvm-trunk-next` on your local private clones to make sure your history matches upstream. If you set `llvm-trunk-next` as the default branch, we can avoid this sort of mixup in the future.

whentojump commented:
> > @chuckwolber Any update on the panic you mentioned in the original post? Does the kernel message look like the below before the panic? If so I think it's expected and can be eliminated by turning off `CONFIG_KUNIT_FAULT_TEST`.
> > ```
> > [    4.111299] ok 26 kunit_device
> > [    4.112655]     KTAP version 1
> > [    4.113192]     # Subtest: kunit_fault
> > [    4.113833]     # module: kunit_test
> > [    4.113837]     1..1
> > [    4.114920] BUG: kernel NULL pointer dereference, address: 0000000000000000
> > ```
> 
> Yes, that is what I saw. I did not test to see if we get a kernel panic without the `llvm-cov` patches applied. If that happens to be the case then we can ignore the problem and omit `CONFIG_KUNIT_FAULT_TEST`. Otherwise, we should probably create an issue and get it resolved.
> 

I can confirm on my side: without instrumentation the panic is also there.

You can verify with the following sets of options:

1. Without the patch at all

```shell
git checkout v6.11-rc3
make LLVM=1 mrproper
make LLVM=1 defconfig
./scripts/config -e CONFIG_9P_FS_POSIX_ACL
./scripts/config -e CONFIG_9P_FS
./scripts/config -e CONFIG_NET_9P_VIRTIO
./scripts/config -e CONFIG_NET_9P
./scripts/config -e CONFIG_PCI
./scripts/config -e CONFIG_VIRTIO_PCI
./scripts/config -e CONFIG_OVERLAY_FS
./scripts/config -e CONFIG_DEBUG_FS
./scripts/config -e CONFIG_CONFIGFS_FS
./scripts/config -e CONFIG_MAGIC_SYSRQ
make LLVM=1 olddefconfig
./scripts/config -e CONFIG_KUNIT
./scripts/config -e CONFIG_KUNIT_ALL_TESTS
make LLVM=1 olddefconfig
make LLVM=1 -j$(nproc)
```

Panic

2. With our patch, but disable instrumentation

```shell
git checkout v1.0-draft
make LLVM=1 mrproper
make LLVM=1 defconfig
./scripts/config -e CONFIG_9P_FS_POSIX_ACL
./scripts/config -e CONFIG_9P_FS
./scripts/config -e CONFIG_NET_9P_VIRTIO
./scripts/config -e CONFIG_NET_9P
./scripts/config -e CONFIG_PCI
./scripts/config -e CONFIG_VIRTIO_PCI
./scripts/config -e CONFIG_OVERLAY_FS
./scripts/config -e CONFIG_DEBUG_FS
./scripts/config -e CONFIG_CONFIGFS_FS
./scripts/config -e CONFIG_MAGIC_SYSRQ
make LLVM=1 olddefconfig
./scripts/config -e CONFIG_KUNIT
./scripts/config -e CONFIG_KUNIT_ALL_TESTS
make LLVM=1 olddefconfig
make LLVM=1 -j$(nproc)
```

Panic

whentojump commented:
I'm fine with any branching strategy. Let me see if I understand your proposed one correctly:

![image](https://github.com/user-attachments/assets/9c263a94-e744-43e2-9ed0-b3f1f64ec665)

If so let's do this way. @chuckwolber After you can confirm it, I will change the default branch of `xlab-uiuc/linux-mcdc` to `llvm-trunk-next`

chuckwolber commented:
> If so let's do this way. @chuckwolber After you can confirm it, I will change the default branch of `xlab-uiuc/linux-mcdc` to `llvm-trunk-next`

Yes, that is correct. I think most collaborative (git based) projects have gravitated to that approach over time.

Merging from the development branch `llvm-trunk-next` to the release branch `llvm-trunk` is totally up to you. Since we are storing patches in a versioned directory, I would be careful about when you merge those to `llvm-trunk`, because we should not change them after that. It would have to be `v2.0` patches or something like that.

whentojump commented:
Default branch has been updated
